runner/src/scheduler.ts

```diff
@@ -127,6 +137,25 @@ export function onError(
   errorHandlers.add(fn);
 }
 
+// Replace the default console hook with a function that accepts context metadata,
+// console method name and the arguments.
+export function onConsole(
+  fn: (
+    metadata: ReturnType<typeof getCharmMetadataFromFrame>,
+    method: ConsoleMethod,
+    args: any[],
+  ) => any[],
+) {
+  consoleHandler = fn;
+}
+
+runtime.addEventListener("console", (e: Event) => {
+  const { method, args } = e as ConsoleEvent;
+  const metadata = getCharmMetadataFromFrame();
```

Comment:
This would rely on the event handler being called synchronously from where the console.* call happened. Even if this is the case, this is a little odd and worth commenting.

See the other comment though

Reply:
Events (`EventTarget`, `Event` and friends) are dispatched synchronously FWIW, the frames from `getCharmMetadataFromFrame()` seems to map correctly